### ISSUE TITLE
Excalidraw whiteboard accessibility issues Reported by Deque Auditor

### ISSUE THREAD

**Issue by msk4862** [excalidraw/excalidraw#7492]
We audited our pages by [Deque](https://www.deque.com/), and one of our pages also included whiteboard integration using Excalidraw, So that also got tested. We want to share the issues that Deque team reported which are related to Excalidraw whitebord, so that we can plan to fix them in upcoming versions.
Excalidraw version used: `0.15.2`

### Issue 1:

**Problem:** IDs of active elements must be unique.

**Source Code:**
```
<input class="ToolIcon_type_radio ToolIcon_size_medium" type="radio" name="editor-current-shape" aria-label="Hand (panning tool) — H" aria-keyshortcuts="h" data-testid="toolbar-hand" id="zIGMdnVRlQIpVGLcllmpL-undefined">
```

**Recommendation to fix:** 
Document has active elements with the same id attribute: `zIGMdnVRlQIpVGLcllmpL-undefined`

### Issue 2:

**Problem:** Hidden or empty element receives focus.

**Source Code:**
```
<button data-event-action="Click" data-event-label="CPSideDrawerImportQuestion" data-cd-interview-id="4550761" data-event-category="CodePair React" type="button" class="c-beGkFe c-beGkFe-cddMLR-variant-secondary c-beGkFe-bwrheh-size-medium hr-inline-flex hr-justify-center hr-align-center hr-body-02 hr-p-y-0.5 hr-p-x-0.75 import-question-section__import-btn" data-hr-focus-item="private">Import Question</button>

<button type="button" class="c-beGkFe c-beGkFe-cddMLR-variant-secondary c-beGkFe-bwrheh-size-medium hr-inline-flex hr-justify-center hr-align-center hr-body-02 hr-p-y-0.5 hr-p-x-0.75 recommended-question-section__button" data-hr-focus-item="private">Recommended Questions</button>
```

**Recommendation to fix:** Document has active elements with the same id attribute: `zIGMdnVRlQIpVGLcllmpL-undefined`
The reading and navigation order MUST be logical and intuitive.

All focusable elements MUST have a visual focus indicator when in focus.

HOW TO FIX: 
Fix this issue by using ONE of the following techniques:
1. If the element should not be visible or focusable, remove the element from the DOM or hide it from all users with CSS such as display:none.
2. If the element is meant to be visible but not interactive, make it visible and remove the focus state.
3. If the element is meant to be visible and interactive, then ensure that it is both focusable and visible to all users.

REFERENCE: 
Deque University: https://dequeuniversity.com/class/input-methods/keyboard-input/tab-order
Deque University: https://dequeuniversity.com/class/input-methods/keyboard-input/visual-focus-indicator

### Issue 3:

**Problem:** The visual focus indicator is missing altogether.

**Source Code:**
```
<label class="ToolIcon ToolIcon__lock ToolIcon_size_medium" title="Keep selected tool active after drawing — Q">...</label>
```

**Element description:** 
Checkbox and radio buttons in the toolbar

**Element location:**
In the "Draw Canvas" region

**Description of the issue:** 
The focus indicator on the above-mentioned elements is missing. Hence it is hard for the users to interact with the controls on the page as there is no visible focus indicator.

**Recommendation to fix:**
RULE : 
All focusable elements MUST have a visual focus indicator when in focus.

Focusable elements SHOULD have enhanced visual focus indicator styles.

HOW TO FIX: 
Common keyboard focus indicator solutions include:
1. A 2px box around the focused element
2. A change in the background color of the focused element
3. The addition of an icon, such as an arrow, next to a menu item
4. The addition of a thick line under or next to the focused element
5. A distinctive change to the text of the component such as making it bold and/or underlined
6. Use the default browser focus indicator. NOTE: Since each browser has its own default focus indicator style, check each of the major browsers (Chrome, Firefox, Edge, IE, Safari) to ensure the default focus indicator is visible as you tab through the page. 

A keyboard focus indicator can take different forms; it does not have to be a boring box. But the goal is to provide a difference in contrast between a component's default and focused states of at least 3 to 1 or another distinctive visual change.

REFERENCE: 
Deque University: https://dequeuniversity.com/class/input-methods/keyboard-input/visual-focus-indicator

![65caf934-900d-11ee-aad2-879b9ea93fab](https://github.com/excalidraw/excalidraw/assets/24875366/5d881e6c-62b0-4783-a4a4-a0feebbc66a2)

### Issue 4:

**Problem:** New window opens automatically without warning.

**Source Code:**
```
<input class="ToolIcon_type_radio ToolIcon_size_medium" type="radio" name="editor-current-shape" aria-label="Insert image" aria-keyshortcuts="9" data-testid="toolbar-image" id="c95xlmXLuOw3U93it11ta-undefined">
```

**Description:**
A window is automatically launched without warning when the user inputs a value in a text field or selects a checkbox or radio button.

**Element description:** 
"Insert image" radio button

**Element location:** 
In the "Shapes" region

**Description of the issue:** 
When the user selects the above-mentioned radio button, a new window to select an image is opened without prior warning or indication. The immediate change in context will confuse the users.

**Recommendation to fix:**
RULE : 
Changing an element's value MUST NOT automatically trigger a change of context, unless the user is adequately advised ahead of time.

HOW TO FIX: 
Fix this issue by using ONE of the following techniques:
1. Ensure that entering text into an edit box, checking a checkbox or radio button, or navigating to one of the options of a select control does not automatically open a new window.
2. Inform users ahead of time of such behavior by methods such as the text label for the UI control or some advisory text placed before the control that cautions the user of this behavior.

REFERENCE: 
Deque University: https://dequeuniversity.com/class/forms/dynamic-custom/changes-context

![88564f0e-9011-11ee-ae41-07027bbedabf](https://github.com/excalidraw/excalidraw/assets/24875366/fc283395-6339-4039-ab73-ba2b3e669167)

### Issue 5:

**Problem:** Function cannot be performed by keyboard alone.

**Source Code:**
```
<input class="ToolIcon_type_radio ToolIcon_size_medium" type="radio" name="editor-current-shape" aria-label="Draw" aria-keyshortcuts="P or 7" data-testid="toolbar-freedraw" id="c95xlmXLuOw3U93it11ta-undefined">
```

**Description:**
There is no way to perform the function using only the keyboard on the same screen or on a qualifying conforming alternate version.

**Element description:** 
Shapes (line, arrow, circle, ellipse, etc)

**Element location:** 
In the "Draw Canvas" region

**Description of the issue:** 
Keyboard users can select the action to perform from the toolbar (Ex: Rectangle, Square, etc) however, they are not able to continue to draw the shapes using the keyboard alone.
Recommendation to fix
RULE : 
Functionality MUST be available using the keyboard, unless the functionality cannot be accomplished in any known way using a keyboard.

HOW TO FIX: 
Fix this issue by using use either a device independent event handler (one that works with both the mouse and the keyboard) or use both mouse dependent and keyboard dependent event handlers.

Device independent event handlers include: onfocus, onblur, onselect, onchange, and onclick (when onclick is used with a keyboard-focusable element).

Bad: <button ondblclick="foo()"">Next</button>
Good: <button onclick="foo()"">Next</button>

REFERENCE: 
Deque University:https://dequeuniversity.com/class/input-methods/keyboard-input/functionality

![2dea56a4-95a3-11ee-9975-0b8afe91b22a](https://github.com/excalidraw/excalidraw/assets/24875366/05611356-fa3d-44a2-9cb5-dad4d744830c)

### Issue 6:

**Problem:** Text content lacks 4.5:1 contrast ratio.

**Source Code:**
```
<div class="HintViewer"><span>To move canvas, hold mouse wheel or spacebar while dragging, or use the hand tool</span></div>
```

**Description:**
The contrast ratio between text and its background is not at least 4.5:1.
Element description: 
"To move Canvas, hold mouse wheel or spacebar while dragging, or use the hand tool" 
"Click to start multiple points, drag for single line"
"Click and drag, release when you are finished"
"Tip: you can also add text by double clicking anywhere within the selection tool"
"Hold Alt to revert the elements marked for deletion"
elements

**Element location:**
In the "Draw Canvas" region
 
**Description of the issue:** 
Contrast Details
Foreground text color: #42464E
Background color: #121212
Contrast Ratio: 1.97 : 1 

The text needs to have a good contrast ratio with a background of 4.5:1 at the minimum. When this is not met, people with color blindness and low vision will find it difficult to read the content and with the above details, we can see that the element is not meeting the minimum threshold of 4.5:1 contrast ratio. 

**Steps to reproduce:**
Select radio buttons to get the different text

**Recommendation to fix:**
RULE : 
Small text (under 18 point regular font or 14 point bold font) MUST have a contrast ratio of at least 4.5 to 1 with the background.

HOW TO FIX: 
Fix this issue by adjusting the text and/or background to increase the contrast to at least 4.5 to 1.

REFERENCE: 
Deque University: https://dequeuniversity.com/class/visual-design/contrast/text-against-background

![9cb15ee0-9010-11ee-a219-8ba07715751c](https://github.com/excalidraw/excalidraw/assets/24875366/4202c1ad-d7ee-4878-a350-2eff51fe3f13) ![9cb15ad0-9010-11ee-874b-ab41b20a68ca](https://github.com/excalidraw/excalidraw/assets/24875366/8b31414b-43bb-44b8-a251-5bf8cc407085) ![9cb14036-9010-11ee-9c9a-7bf2453d7a33](https://github.com/excalidraw/excalidraw/assets/24875366/d5b24c73-1bca-466d-b7d4-9a86aa56bdd5) ![9caf5d20-9010-11ee-810c-876d0778e937](https://github.com/excalidraw/excalidraw/assets/24875366/07599328-2d8a-409b-9b89-1c68e005a0b9) ![9caf578a-9010-11ee-8c74-3b4830a0a75f](https://github.com/excalidraw/excalidraw/assets/24875366/ca761944-4dda-4bd4-98ce-b6d4fd0f7910)

### Issue 7:

**Problem:** State of active component lacks 3 to 1 contrast ratio.

**Source Code:**
```
<input class="ToolIcon_type_radio ToolIcon_size_medium" type="radio" name="editor-current-shape" aria-label="Selection" aria-keyshortcuts="V or 1" data-testid="toolbar-selection" id="c95xlmXLuOw3U93it11ta-undefined" checked="">

<input class="ToolIcon_type_checkbox" type="checkbox" aria-label="Keep selected tool active after drawing" data-testid="toolbar-lock">
```

**Description:**
The state of an active component lacks 3 to 1 contrast ratio.
Element description: 
"Keep selected tool active after drawing" checkbox
"Selection" radio button

**Element location:** 
In the "Draw Canvas" region

**Description of the issue:** 
Contrast Details
Selected state color: #1D1D20
Adjacent color: #27272B
Contrast Ratio: 1.13 : 1 

When the color contrast ratio of an interactive element's state with its background is less than 3:1, it becomes difficult for a color blind or low vision user to understand different states of the element. 

**Recommendation to fix:**
RULE : 
The visual state of an active user interface component MUST have sufficient contrast of 3 to 1 with the adjacent background. Exceptions exist.

HOW TO FIX: 
Fix this issue by adjusting the visual focus indicator or state indicator (e.g. selected, checked, pressed, etc.) of the user interface component and/or background to increase the contrast with either the inner or outer adjacent background to at least 3 to 1.

REFERENCE: 
Deque University: https://dequeuniversity.com/class/visual-design/contrast/non-text-contrast
WCAG Understanding document: https://www.w3.org/WAI/WCAG22/Understanding/non-text-contrast

![eb962ab8-900b-11ee-a18f-5bef7a05d7ba](https://github.com/excalidraw/excalidraw/assets/24875366/bc5c69c7-9d45-4451-a594-d2953d3d6c08)

### Issue 8:

**Problem:** Buttons must have discernible text.

**Source Code:**
```
<button data-prevent-outside-click="true" class="dropdown-menu-button  zen-mode-transition transition-left" type="button" data-testid="dropdown-menu-button">
```

**Description:**
Buttons must have discernible text

**Recommendation to fix:**

Fix any of the following:
  Element does not have inner text that is visible to screen readers
  aria-label attribute does not exist or is empty
  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
  Element has no title attribute
  Element's default semantics were not overridden with role="none" or role="presentation"

**Comment by hannahbergam** [comment]
Hello! Code.org is also working on an a11y audit and has found these issues and more. Is there any plan to make progress on this list?

**Comment by Lollihopss** [comment]
Hi! I'm a computer science student who recently ran a WAVE accessibility evaluation on Excalidraw and noticed several contrast and focus indicator issues. Before opening a new issue, I searched existing tickets and found this one, which documents the same problems I found. I'd love to start with Issue 6 (HintViewer text contrast ratio of 1.97:1, failing WCAG AA minimum of 4.5:1) and potentially tackle additional issues from this list depending on complexity. Could this be assigned to me, or should I just proceed with a PR? Happy to discuss the approach beforehand!
